src/eos-updater-fetch.c

```diff
@@ -744,9 +758,18 @@ handle_fetch (EosUpdater            *updater,
       return TRUE;
     }
 
+  fetch_data = g_new0 (FetchData, 1);
+  fetch_data->update_id = g_strdup (eos_updater_get_update_id (updater));
+  fetch_data->update_refspec = g_strdup (eos_updater_get_update_refspec (updater));
+  fetch_data->progress = ostree_async_progress_new_and_connect (update_progress, updater);
+
+  /* FIXME: Passing the EosUpdaterData to the worker thread is not thread safe.
+   * See: https://phabricator.endlessm.com/T15923 */
+  fetch_data->data = data;
+
   eos_updater_clear_error (updater, EOS_UPDATER_STATE_FETCHING);
-  task = g_task_new (updater, NULL, content_fetch_finished, user_data);
-  g_task_set_task_data (task, user_data, NULL);
+  task = g_task_new (updater, NULL, content_fetch_finished, NULL);
+  g_task_set_task_data (task, g_steal_pointer (&fetch_data), (GDestroyNotify) fetch_data_free);
```

Comment:
As an added benefit, if you have a `fetch_data_new` here, then there's no need for `g_steal_pointer` etc, just pass the result directly to `g_task_set_task_data`